austation/code/modules/popes-xenoarchaeology/xenoartifact_labeler.dm

```diff
@@ -0,0 +1,214 @@
+    if(action == "print_traits")
+        create_label(sticker_name)
+
+    if(action == "change_print_name")
+        sticker_name = params["name"]
+
+    trait_toggle(action, "activator", activator_traits, activator)
+    trait_toggle(action, "minor", minor_traits, minor_trait)
+    trait_toggle(action, "major", major_traits, major_trait)
+    trait_toggle(action, "malfunction", malfunction_list, malfunction)
+
+    . = TRUE
+    update_icon()
+
+/obj/item/xenoartifact_labeler/proc/get_trait_list_desc(list/traits, trait_type)//Get a list of all the specified trait types names, actually
+    for(var/T in typesof(trait_type))
+        var/datum/xenoartifact_trait/X = new T
+        if(X.desc && !(X.desc in traits) && !(X.label_name))
+            traits += list(X.desc)
+        else if(X.label_name && !(X.label_name in traits)) //For cases where the trait doesn't have a desc or is tool cool to use one
+            traits += list(X.label_name)
+    return traits
+
+/obj/item/xenoartifact_labeler/proc/look_for(list/place, culprit) //This isn't really needed but, It's easier to use as a function. What does this even do?
+    for(var/X in place)
+        if(X == culprit)
+            return TRUE
+    return FALSE
+
+/obj/item/xenoartifact_labeler/afterattack(atom/target, mob/user)
+    ..()
+    var/obj/item/xenoartifact_label/P = create_label(sticker_name)
+    if(!P.afterattack(target, user)) //In the circumstance the sticker fails, usually means you're doing something you shouldn't be
+        qdel(P)
+
+/obj/item/xenoartifact_labeler/proc/create_label(new_name)
+    var/obj/item/xenoartifact_label/P = new /obj/item/xenoartifact_label/(get_turf(src))
+    if(new_name)
+        P.name = new_name
+        P.set_name = TRUE
+    P.trait_list = sticker_traits
+    P.info = activator+minor_trait+major_trait
+    return P
+
+/obj/item/xenoartifact_labeler/proc/trait_toggle(action, toggle_type, var/list/trait_list, var/list/active_trait_list)
+    var/datum/xenoartifact_trait/description_holder
+    var/new_trait
+    for(var/T in trait_list)
+        new_trait = desc2datum(T)
+        description_holder = new new_trait
+        if(action == "assign_[toggle_type]_[T]")
+            if(!look_for(active_trait_list, T))
+                active_trait_list += list(T)
+                info_list += description_holder.label_desc
+                sticker_traits += new_trait
+            else
+                active_trait_list -= list(T)
+                info_list -= description_holder.label_desc
+                sticker_traits -= new_trait
+
+/obj/item/xenoartifact_labeler/proc/desc2datum(udesc) //This is just a hacky way of getting the info from a datum using its desc becuase I wrote this last and it's not heartbreaking
+    for(var/T in typesof(/datum/xenoartifact_trait))
+        var/datum/xenoartifact_trait/X = new T
+        if((udesc == X.desc)||(udesc == X.label_name))
+            return T
+    return "[udesc]: There's no known information on [udesc]!."
```

Comment:
why 😢

Reply:
I tacked this on last after I wrote the whole system, because I didn't originally intend descriptions to exist. So, unfortunately very little hard data is passed other than strings. 
I can revisit this if it's absolutely repulsive.